src/gc.h

```diff
@@ -175,76 +175,99 @@ typedef struct {
     uint16_t fl_end_offset;   // offset of last free object in this page
     uint16_t thread_n;        // thread id of the heap that owns this page
     char *data;
+    struct _jl_gc_pagemeta_t *next;
 } jl_gc_pagemeta_t;
 
+typedef struct {
+    jl_mutex_t lock;
+    jl_gc_pagemeta_t *page_metadata_back;
+} jl_gc_global_page_pool_t;
+
+extern jl_gc_global_page_pool_t global_page_pool;
+
+#define GC_ALLOC_MAP_LG2 ((sizeof(void*) * 8 - GC_PAGE_LG2) / 2)
+
+typedef struct {
+    uint8_t alloc[1ull << GC_ALLOC_MAP_LG2];
+} jl_gc_alloc_map_inner_t;
+
+typedef struct {
+    jl_mutex_t lock;
+    jl_gc_alloc_map_inner_t *alloc[1ull << GC_ALLOC_MAP_LG2];
+} jl_gc_alloc_map_t;
+
+extern jl_gc_alloc_map_t alloc_map;
+
+STATIC_INLINE size_t gc_alloc_map_inner_idx(void *_data) JL_NOTSAFEPOINT
+{
+    return (size_t)(((uintptr_t)(_data) << GC_ALLOC_MAP_LG2) >> (GC_ALLOC_MAP_LG2 + GC_PAGE_LG2));
+}
+
+STATIC_INLINE size_t gc_alloc_map_idx(void *_data) JL_NOTSAFEPOINT
+{
+    return (size_t)(uintptr_t)(_data) >> (GC_ALLOC_MAP_LG2 + GC_PAGE_LG2);
+}
+
+STATIC_INLINE uint8_t gc_alloc_map_is_set(char *_data) JL_NOTSAFEPOINT
+{
+    jl_gc_alloc_map_inner_t *inner = alloc_map.alloc[gc_alloc_map_idx(_data)];
+    if (inner != NULL) {
+        return inner->alloc[gc_alloc_map_inner_idx(_data)];
+    }
+    return 0;
+}
+
+STATIC_INLINE void gc_alloc_map_maybe_create(char *_data) JL_NOTSAFEPOINT
+{
+    jl_gc_alloc_map_inner_t *inner = alloc_map.alloc[gc_alloc_map_idx(_data)];
+    if (inner == NULL) {
+        inner = (jl_gc_alloc_map_inner_t*)calloc_s(sizeof(jl_gc_alloc_map_inner_t));
+        alloc_map.alloc[gc_alloc_map_idx(_data)] = inner;
+    }
+}
+
+STATIC_INLINE void gc_alloc_map_set(char *_data, uint8_t v) JL_NOTSAFEPOINT
+{
+    jl_gc_alloc_map_inner_t *inner = alloc_map.alloc[gc_alloc_map_idx(_data)];
+    assert(inner != NULL);
+    inner->alloc[gc_alloc_map_inner_idx(_data)] = v;
+}
 // Page layout:
 //  Newpage freelist: sizeof(void*)
```

Comment:
```suggestion
```

?

Reply:
I'm still keeping the newpage freelist pointer in this implementation.

Reply:
That will break our sizeclass computational arithmetic, forcing us to recompute some of those in this PR. Can we avoid doing this?

Reply:
Had misunderstood your point about the sizeclass. Yes, I think this should be possible.